### PR TITLE
torch2.0.1 custom auto wrap

### PR DESCRIPTION
**Pull request by vchiley**
# What does this PR do?

Creates a torch2.0.1 version of https://github.com/mosaicml/composer/pull/2006 by monkey patching fsdp's \_\_init__ fn where we call the custom version of `_auto_wrap`.

Known issue: Does not work with `use_orig_params: true` + `model: init_device: meta`

cc @abhi-mosaic btw I'm getting rid of the MosaicFullyShardedDataParallel cls and just monkey patching the appropriate member methods instead.

# What issue(s) does this change relate to?

https://github.com/mosaicml/composer/pull/2006 didn't work with torch2.0.1

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by mvpatel2000 — COMMENTED** [review]
Is this ready for review? I see tagged but draft. Also does this support backwards compatibility?

**Comment by vchiley** [comment]
@mvpatel2000 I want to spend some time on this first:
> Known issue: Does not work with `use_orig_params: true` + `model: init_device: meta`

File: composer/trainer/mosaic_fsdp.py [hunk 1]
```diff
@@ -321,4 +329,292 @@ def _auto_wrap(
...
+        ignored_modules: Set[nn.Module],
+        ignored_params: Set[nn.Parameter],
+        process_group_cache: Dict[Tuple[int], Any],
+        only_wrap_children: bool = False,
+        **kwargs: Any,
+    ) -> Tuple[nn.Module, int]:
+        """Updates FSDPs _recursive_wrap to enable module_kwargs and custom process_group cache.
+
+        modified version of
+        https://github.com/pytorch/pytorch/blob/96ca226a7332be0d8f3d6159d0c797e032ab0721/torch/distributed/fsdp/wrap.py
```
Comment (bcui19) [inline composer/trainer/mosaic_fsdp.py:367]:
```suggestion
        https://github.com/pytorch/pytorch/blob/96ca226a7332be0d8f3d6159d0c797e032ab0721/torch/distributed/fsdp/wrap.py#L320
```

**Review by bcui19 — COMMENTED** [review]
Mostly LGTM, just minor questions, nits

File: composer/trainer/mosaic_fsdp.py [hunk 2]
```diff
@@ -321,4 +329,292 @@ def _auto_wrap(
...
+                    process_group_cache=process_group_cache,
+                    **kwargs,
+                )
+                setattr(module, name, wrapped_child)
+                # Keep track of how many parameters have been wrapped
+                total_wrapped_numel += num_wrapped_params
+            # decide if we need to wrap the current module,
+            # since the left over parameters exceed the number of params to wrap
+            remainder = nonwrapped_numel - total_wrapped_numel
+            module_kwargs = auto_wrap_policy(module=module, recurse=False, nonwrapped_numel=remainder)
```
[the remark this answers is not in the document]
Comment (vchiley) [reply composer/trainer/mosaic_fsdp.py:430]: `auto_wrap_policy` returning a dict of per fsdp module custom fsdp kwargs is the point of this entire PR 😄 
This allows us to do something like this which restricts the gpu sets which fsdp sync's params / gradients.

[the remark this answers is not in the document]
Comment (bcui19) [reply composer/trainer/mosaic_fsdp.py:433]: Ah right sorry, thank you for the refresher :)

File: composer/trainer/mosaic_fsdp.py [hunk 3]
```diff
@@ -175,112 +173,122 @@ def get_process_group(pg, process_group_cache=None):
...
+                    **kwargs,
+                )
+                setattr(module, name, wrapped_child)
+                # Keep track of how many parameters have been wrapped
+                total_wrapped_params += num_wrapped_params
+            # decide if we need to wrap the current module,
+            # since the left over parameters exceed the number of params to wrap
+            remainder = num_params - total_wrapped_params
+            module_kwargs = auto_wrap_policy(module=module, recurse=False, unwrapped_params=remainder)
+            if not only_wrap_children and module_kwargs:
```
[the remark this answers is not in the document]
Comment (vchiley) [reply composer/trainer/mosaic_fsdp.py:282]: valid
keeling both `_custom_recursive_wrap_t1p13p1` and `_custom_recursive_wrap_t2p0p1` but creating
```
def set_custom_fsdp_module_kwargs(module_kwargs: Dict) -> Dict:
```
since its the same in both versions
